### PR TITLE
[nexus] garbage collect orphaned FM sitreps

### PR DESCRIPTION
**Pull request by hawkw**
When a Nexus attempts to commit a new fault management situation report to the sitrep history but fails to do so because another sitrep with the same parent has already been inserted, that sitrep is said to be _orphaned_. Records pertaining to it are left behind in the database, but it will not be accessed by the rest of the system. Thus, we must occasionally garbage-collect such sitreps. This branch adds a background task for doing so.

Depends on #9320

### CODE REVIEW

File: dev-tools/omdb/src/bin/omdb/db/sitrep.rs [hunk 1]
```diff
@@ -138,53 +138,58 @@ pub(super) async fn cmd_db_sitrep_history(
...
+            }
+        };
+        let (comment, time_created) =
+            match datastore.fm_sitrep_metadata_read(&opctx, v.id).await {
+                Ok(s) => (s.comment, Some(s.time_created)),
+                Err(e) => {
+                    eprintln!(
+                        "failed to get fetch metadata for sitrep {} (v{}): {e}",
+                        v.id, v.version
+                    );
```
Comment (smklein) [inline dev-tools/omdb/src/bin/omdb/db/sitrep.rs]: Would this be the expected case if we did a partial insert of sitrep data, but the "final sitrep" was not inserted?

Or are we saying: "if any records exist correlated with a sitrep, it is guaranteed to have an entry in the  `fm_sitrep` table"?

File: nexus/src/app/background/tasks/fm_sitrep_gc.rs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    async fn actually_activate(&mut self, opctx: &OpContext) -> Status {
+        let mut status = Status::default();
+        let mut total_found = 0;
+        let mut total_deleted = 0;
+        let mut paginator =
+            Paginator::new(SQL_BATCH_SIZE, PaginationOrder::Descending);
+
+        while let Some(p) = paginator.next() {
```
Comment (smklein) [inline nexus/src/app/background/tasks/fm_sitrep_gc.rs:54]: My read on this is:

- We only consider orphans "relative to a version"
- We only consider versions that are returned from `fm_sitrep_version_list`

Does this mean that if any rows are removed from `fm_sitrep_history`, we won't be able to clean up the corresponding orphans which might exist for that version?

Do we have any protection against this leak? If no: are we documenting that this is a critical constraint we need to satisfy before deleting any `fm_sitrep_history` rows

File: nexus/src/app/background/tasks/fm_sitrep_gc.rs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+                    status.errors.push(format!(
+                        "failed to list sitrep versions: {err}",
+                    ));
+                    break;
+                }
+            };
+
+            paginator = p.found_batch(&versions, &|v| SqlU32::from(v.version));
+
+            for v in versions {
```
Comment (smklein) [inline nexus/src/app/background/tasks/fm_sitrep_gc.rs]: (Related to my earlier comment) I think there's some complexity in identifying "orphans by version", since it makes it unsafe to delete any part of the version history while orphans might remain.

It might be possible to do orphan deletion in a more version-independent way:

- Lookup all sitreps that are *not* referenced by `fm_sitrep_history`
- Delete all these sitreps, unless `parent_sitrep_id` is the current active sitrep

Basically:

```sql
WITH current_sitrep_id AS (
    SELECT sitrep_id
    FROM omicron.public.fm_sitrep_history
    ORDER BY version DESC
    LIMIT 1
),
-- Scan through `fm_sitreps` to avoid full table scans
batch AS (
    SELECT s.id, s.parent_sitrep_id
    FROM omicron.public.fm_sitrep s
    WHERE s.id > $1
    ORDER BY s.id
    LIMIT $2
)
SELECT FROM omicron.public.fm_sitrep
WHERE id IN (
    SELECT b.id
    FROM batch b
    -- Lookup the sitrep in the history
    LEFT JOIN omicron.public.fm_sitrep_history h ON h.sitrep_id = b.id
    -- Find sitreps missing from history
    WHERE h.sitrep_id IS NULL
      -- ... where the sitrep cannot be made active
      AND (b.parent_sitrep_id IS NULL 
           OR b.parent_sitrep_id != (SELECT sitrep_id FROM current_sitrep_id))
)
RETURNING id;
```

File: nexus/src/app/background/tasks/fm_sitrep_gc.rs (hunk 2 above)
Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs:54]: I was about to add: "As long as we check for no orphans before removing an old row from `fm_sitrep_history`, this will be safe", but I think that could still cause orphaned sitreps to be leaked if:

- We performed history cleanup for an old row, and removed all orphans
- Concurrently, a slow-old-nexus inserted a sitrep referencing that very old parent (obviously it would not be able to insert into `fm_sitrep_history`)

In this case, a sitrep would be inserted, referencing an old parent, and we'd never scan the necessary version to delete such a sitrep

Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs:54]: Also: If we "just don't delete versions", then the background task will linearly increase the number of queries it needs to perform forever, which would be bad.

File: dev-tools/omdb/src/bin/omdb/db/sitrep.rs (hunk 1 above)
Comment (smklein) [reply dev-tools/omdb/src/bin/omdb/db/sitrep.rs]: Edit: Seems like you're relying on that ordering, as I'm reading below.

File: nexus/src/app/background/tasks/fm_sitrep_gc.rs (hunk 3 above)
Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Ah, yeah... I had done it based on entries in the version table as a way to avoid doing a full table scan, but I think this approach would also work.

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Hmm, so, I was originally thinking I would do this using the current sitrep ID from the `sitrep_loader` watch channel, rather than as a CTE, but I realize that actually doesn't work. The current sitrep in the database may have moved, and if a "stale" current sitrep ID was used for this query, it would return children of the current sitrep which are in progress. I'm not sure if the CTE is safe either: would this ensure that the current sitrep is locked while we are executing the `SELECT`, or could it have moved forwards in the midst of this query?

Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Unless I wrote this proposed CTE incorrectly, my intent was that:

- Every sitrep in history
- Every sitrep which **could** be added to history

Are never going to be returned from the SELECT. In fact, we'd only return the opposite: sitreps which are not in history, and which cannot be in history.

As a result: this **should** be safe from concurrent insertions - current / possible future sitreps shouldn't be returnable from it.

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: What I wasn't sure about was whether the `fm_sitrep_history` table would be locked as long as the query is executing. If it is, then this should be fine. But if a new current sitrep could be added after we observe the current ID, then I _think_ we would incorrectly observe children of the new current sitrep as eligible for GC.

Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: As currently written, this is a CTE, so it should do lock `fm_sitrep_history` (basically, whatever value it thinks is "latest" should remain valid for the duration of the statement).

If we run this CTE while...
- ... other Nexuses are adding new sitreps, which **could** become active, but are not yet: they won't be eligible for removal, because the CTE explicitly ignores them
- ... other Nexuses are concurrently adding **old** sitreps (using an old parent), those **would** be eligible for removal. But that's okay, it's sorta indistinguishable from them adding old sitreps "a long time ago".
- ... other Nexuses are adding a new row to `fm_sitrep_history`, this CTE will run either before that row addition (and explicitly ignore the possible-new sitrep) or after that row addition (and it will explicitly ignore an `fm_sitrep` with an entry in `fm_sitrep_history`).

Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: To answer your question more generally: I believe that a single-statement CTE is "implicitly a transaction" from the perspective of CockroachDB. So, from an ACID point-of-view:

```
WITH
FOO AS (...)
BAR AS (...)
SELECT ...
```

Should roughly have the same semantics as:

```
BEGIN
let foo = SELECT (...);
let bar = SELECT(...);
SELECT ...;
COMMIT;
```

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Okay, cool, thank you. I wasn't totally sure what the locking semantics were here, so thanks for explaining. In that case, this seems like the right thing.

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Ta-dah: 642854f56039e848a44d5d559910b3df928cb5fe (man, writing CTEs in Diesel *really* sucks)

Comment (smklein) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: (I still need to re-review the rest of this, but:)

I recommend looking at `nexus/db-queries/src/db/raw_query_builder.rs` -- there's a struct called `QueryBuilder` which I added, and which we're using in a few spots in Omicron.

Basically: I agree, the Diesel mechanism to write CTEs is incredibly painful. It **sorta** integrates with the existing Diesel type system, and allows passing named columns/tables, nested ASTs via `walk_ast`, but IMO the syntax is complex enough and **already involves pushing raw SQL strings**, so the extra complexity becomes a bit of a burden. I have the opinion that: "if you need to have a comment clarifying what your SQL query should be, right next to where you actually write your SQL query, something has gone wrong".

`QueryBuilder` tries to limit your input to:
- Raw SQL as a string (with some validations to prevent SQL injection)
- Bind parameters (which still use Diesel `sql_types`, but just that -- no "diesel query" involvement)

We also have `expectorate_query_contents` to make it easy to validate "the query is valid SQL" before any other testing.

If I'm writing a simple, non-CTE diesel query: I usually reach for Diesel's query system first. The types are validated, it's almost always valid SQL, no raw strings, etc.

However, if I'm writing CTEs: The diesel support is so bad (mostly: using diesel's types for columns, table names, subqueries, etc, ends up **getting in the way** more than it helps) - so I just go for the `QueryBuilder` approach.

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Oh, that's nice, I think I overlooked the existence of the more unchecked `QueryBuilder` API. I think that could definitely have helped me avoid a lot of pain from accidental SQL syntax errors from splitting the query up into a bunch of tiny `push_sql` calls. I'll see about rewriting this to use that which should hopefully make future modifications less painful.

Comment (hawkw) [reply nexus/src/app/background/tasks/fm_sitrep_gc.rs]: Yeah, this is much nicer (see 08ce99b2411cb19d2d02cd3be35549e2de32f8be). I'll probably give the other CTE in this module the same treatment in a follow-up PR.

File: nexus/db-queries/src/db/datastore/fm.rs [hunk 4]
```diff
@@ -194,6 +202,168 @@ impl DataStore {
...
+            "SELECT id \
+            FROM omicron.public.fm_sitrep \
+            WHERE id IN ( \
+                SELECT b.id \
+                FROM batch b \
+                LEFT JOIN omicron.public.fm_sitrep_history h \
+                    ON h.sitrep_id = b.id \
+                WHERE
+                    h.sitrep_id IS NULL \
+                AND (
```
Comment (smklein) [inline nexus/db-queries/src/db/datastore/fm.rs]: nitpick; idk if you care about the linebreak after "WHERE" / "AND (".

The "expectorate" tests use crdb's sqlfmt, so it shouldn't make a difference in the output

File: nexus/db-queries/src/db/datastore/fm.rs [hunk 5]
```diff
@@ -521,6 +695,102 @@ mod tests {
         let conn = pool.claim().await.unwrap();
 
         let query = InsertSitrepVersionQuery { sitrep_id: SitrepUuid::nil() };
+
+        // Before trying to explain the query, elt's start by making sure it's
```
Comment (smklein) [inline nexus/db-queries/src/db/datastore/fm.rs]:
```suggestion
        // Before trying to explain the query, let's start by making sure it's
```

File: nexus/db-queries/tests/output/sitrep_list_orphans_no_marker.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  id
+  IN (
+      SELECT
+        b.id
+      FROM
+        batch AS b LEFT JOIN omicron.public.fm_sitrep_history AS h ON h.sitrep_id = b.id
+      WHERE
+        h.sitrep_id IS NULL
+        AND (
+            b.parent_sitrep_id IS NULL
```
Comment (smklein) [inline nexus/db-queries/tests/output/sitrep_list_orphans_no_marker.sql]: Now that I'm looking at this query more closely, I think there *miiiiight* be one edge case here: the case where we have no history.

In this scenario:
- `h.sitrep_id IS NULL` because a new sitrep won't appear in history
- `b.parent_sitrep IS NULL` because the batch of sitrep(s) won't have a parent

So it's technically possible that:
- If we create a new sitrep, but haven't yet inserted it into history...
- ... and this background task runs concurrently...
- ... we'll delete the first sitrep we're trying to insert.

This shouldn't happen afterwards. I'm not sure it even matters for this first sitrep. but figured I'd call it out?

We could probably avoid this by changing this check to:

`(b.parent_sitrep_id IS NULL AND current_sitrep_id.sitrep_id IS NOT NULL)`

**Review by smklein — APPROVED** [review]

File: nexus/db-queries/tests/output/sitrep_list_orphans_no_marker.sql (hunk 6 above)
Comment (hawkw) [reply nexus/db-queries/tests/output/sitrep_list_orphans_no_marker.sql]: Ah, good catch, I think you're right, we should handle that too. I'll make that change.

Comment (hawkw) [reply nexus/db-queries/tests/output/sitrep_list_orphans_no_marker.sql]: I think this actually has to be:
```sql
b.parent_sitrep_id IS NULL AND (SELECT sitrep_id FROM current_sitrep_id) IS NOT NULL
```
but otherwise, yeah.

File: dev-tools/omdb/src/bin/omdb/db/sitrep.rs (hunk 1 above)
Comment (hawkw) [reply dev-tools/omdb/src/bin/omdb/db/sitrep.rs]: bf33ebf1498c687c85eb584cf04933b47fa23ff4 added a quick comment describing this expectation.
